File: 0739-daily-temperatures/0739-daily-temperatures.ts
function dailyTemperatures(temperatures: number[]): number[] {
        // Init res array
        const res: number[] = new Array(temperatures.length).fill(0);
        // Init stack of [temp, index]
        const stack: [number,number][] = [];
        // Iterate through the temperatures array
            // While stack is not empty and the current temperature is greater than the top of the stack:
            //      pop the top of the stack
            //      add the index difference to res
            // Append the current temperature and index to the stack
        for (const [index,temp] of temperatures.entries()) {
            while(stack.length !== 0 && temp > stack[stack.length -1][0]) {
                const [stackTemp, stackIndex] = stack.pop();
                res[stackIndex] = index - stackIndex;
            }
            stack.push([temp,index]);
        }   
      
        // Return the result array
        return res;
    
};

File: 0739-daily-temperatures/0739-daily-temperatures.py
class Solution:
    # Time complexity : O(n)
    # Space complexity : O(n)
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        res = [0] * len(temperatures)
        stack = []
        
        for index,value in enumerate(temperatures):

                
            while stack and value > stack[-1][1]:
                curr_index,curr_temperature = stack.pop()
                res[curr_index] = index - curr_index
            
            
            stack.append([index,value])
            # print(f"stack: {stack}")
        return res
        
        
        
        
        
       
    
    
    
    
    
    
    
    
    
        # # init res array
        # res = [0] * len(temperatures)
        # # init stack of [temp,index]
        # stack = []
        # # iterate index, temp of temperatures:
        # #   while stack is not empty and temp > temp top of stack:
        # #       pop from our stack
        # #       add to res array
        # #   append current index and temp to stack
        # for index,temp in enumerate(temperatures):
        #     while stack and temp > stack[-1][0]:
        #         stack_temp, stack_index = stack.pop()
        #         print(stack_index)
        #         res[stack_index] = index - stack_index
        #     stack.append([temp,index])
        # # return res 
        # # no need to run through res to assign 0 because initially we alrd did
        # return res